### PR TITLE
add code to measure memory usage

### PR DESCRIPTION
**Pull request by pmnoxx**
draft only

### CODE REVIEW

File: chain/chain/src/test_utils.rs [hunk 1]
```diff
@@ -44,18 +45,29 @@ use crate::types::{ApplyTransactionResult, BlockHeaderInfo, ChainGenesis};
 use crate::{BlockHeader, DoomslugThresholdMode, RuntimeAdapter};
 
 #[derive(
-    BorshSerialize, BorshDeserialize, Serialize, Hash, PartialEq, Eq, Ord, PartialOrd, Clone, Debug,
+    BorshSerialize,
```
Comment (bowenwang1996) [inline chain/chain/src/test_utils.rs:48]: we don't need to add this to test-only structs

File: Cargo.toml [hunk 2]
```diff
@@ -75,13 +77,16 @@ neard = { path = "./neard"}
 
 testlib = { path = "./test-utils/testlib" }
 
+memory-tracker = { path = "./memory-tracker" }
+
 [profile.release]
-lto = true        # Enable full link-time optimization.
-codegen-units = 1 # Use only 1 codegen-unit to enable full optimizations.
+lto = false        # Enable full link-time optimization.
+codegen-units = 256 # Use only 1 codegen-unit to enable full optimizations.
```
Comment (bowenwang1996) [inline Cargo.toml:84]: why do we change these two configs here?

File: Cargo.toml [hunk 3]
```diff
@@ -75,13 +77,16 @@ neard = { path = "./neard"}
...
 overflow-checks = true
+# jemallocator = { version = "0.3.0", features = ["unprefixed_malloc_on_supported_platforms", "profiling"] }
 
 [profile.bench]
-lto = true
+lto = false
```
Comment (bowenwang1996) [inline Cargo.toml:89]: also here

File: chain/chain/src/chain.rs [hunk 4]
```diff
@@ -173,6 +176,8 @@ impl OrphanBlockPool {
 
 /// Facade to the blockchain block processing and storage.
 /// Provides current view on the state according to the chain state.
+/// #[
```
Comment (bowenwang1996) [inline chain/chain/src/chain.rs:180]: ?

**Review by bowenwang1996 — COMMENTED** [review]
Looks good. I think after we validated that it works, some of the next steps I see are:
1) put memory tracker related code under a compile-time flag
2) extract deepsize out (probably to its own repo)

File: Cargo.toml (hunk 2 above)
Comment (pmnoxx) [reply Cargo.toml:84]: I disabled those settings to decrease build time from 5m to 17s on my local machine. Mozilla was dealing with similar issue for Firefox, they ended up disabling lto for dev machines, but they enabled it only for release builds:
https://bugzilla.mozilla.org/show_bug.cgi?id=1386371
here is their PR - https://hg.mozilla.org/integration/autoland/rev/d75f07915a48
I think we should do something similar.

File: Cargo.toml (hunk 3 above)
Comment (pmnoxx) [reply Cargo.toml:89]: fixed

File: chain/chain/src/chain.rs (hunk 4 above)
Comment (pmnoxx) [reply chain/chain/src/chain.rs]: fixed

File: chain/chain/src/test_utils.rs (hunk 1 above)
Comment (pmnoxx) [reply chain/chain/src/test_utils.rs:48]: We need it for KeyValueRuntime in order for it to implement RuntimeAdater trait.

If needed I can change the code, so we don't do it for structures used by KeyValueRuntime,

**Comment by pmnoxx** [comment]
> Looks good. I think after we validated that it works, some of the next steps I see are:
> 
> 1. put memory tracker related code under a compile-time flag
I also need to change how the data is presented to be more readable.
> 2. extract deepsize out (probably to its own repo)
Is there a way to extract a package into it's own repo, without making crate public on crates.io.
Or do you think it's fine to release a new package for crates.io?


I'm planning to add more things to this PR after I get initial results about the memory usage. It really depends on where the memory leak is located.

**Comment by bowenwang1996** [comment]
@pmnoxx why did you close this PR?

**Comment by pmnoxx** [comment]
I merged https://github.com/near/nearcore/pull/3747
In that change we added memory profiling, as well as printing memory usage per thread.

We already moved  memory tracker related code to it's own repo.

I'll keep this issue open. I forgot I still need to create a change for deepsize package.

**Comment by bowenwang1996** [comment]
@pmnoxx I see. Feel free to close this PR then

**Pull request closed without merging**